### PR TITLE
Add a profiler for engine replays

### PR DESCRIPTION
**Pull request by cocreature**
This is the sibling of the benchmark infrastructure for ledger
replays.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by cocreature** [comment]
@remyhaemmerle-da do you have a great idea for a test here? I can probably setup some integration test but that is a bit of a pain. I guess what really needs testing is the adapter but that seems slightly out of scope given that I’m not touching that at all.

**Comment by remyhaemmerle-da** [comment]
@cocreature wrote:
> @remyhaemmerle-da do you have a great idea for a test here? I can probably setup some integration test but that is a bit of a pain. I guess what really needs testing is the adapter but that seems slightly out of scope given that I’m not touching that at all.

Can you use the `ledger/participant-state/kvutils/reference-ledger-export` ? I assume there is at least one transaction to replay inside.

**Review by remyhaemmerle-da — APPROVED** [review]
Looks Nice. Thanks.

**Comment by cocreature** [comment]
> Can you use the `ledger/participant-state/kvutils/reference-ledger-export` ? I assume there is at least one transaction to replay inside.

Added a basic `sh_test` which runs the profiler on this. Not great but at least better than nothing.

**Comment by remyhaemmerle-da** [comment]
> Added a basic `sh_test` which runs the profiler on this. Not great but at least better than nothing.

:+1:

File: ledger/participant-state/kvutils/tools/engine-replay/src/profile/scala/ledger/participant/state/kvutils/tools/engine/replay/ReplayProfile.scala [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+  private val parser: OptionParser[Config] = new scopt.OptionParser[Config]("replay-profile") {
+    opt[String]("choice")
+      .action((x, c) => c.copy(choiceName = x))
+      .text("Choice name in the form Module:Template:Choice")
+      .required()
+    opt[Int]("choice-index")
+      .action((x, c) => c.copy(choiceIndex = Some(x)))
+      .text(
+        "If the choice got exercised more than once, you can use this to select the nth occurence of this choice"
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/tools/engine-replay/src/profile/scala/ledger/participant/state/kvutils/tools/engine/replay/ReplayProfile.scala:39]: Typo: `occurrence`

**Pull request merged**